mmhuman3d/core/evaluation/eval_utils.py

```diff
@@ -0,0 +1,199 @@
+    """Calculate the mean per-joint position error (MPJPE) and the error after
+    rigid alignment with the ground truth (P-MPJPE).
+    batch_size: N
+    num_keypoints: K
+    keypoint_dims: C
+    Args:
+        pred (np.ndarray[N, K, C]): Predicted keypoint location.
+        gt (np.ndarray[N, K, C]): Groundtruth keypoint location.
+        mask (np.ndarray[N, K]): Visibility of the target. False for invisible
+            joints, and True for visible. Invisible joints will be ignored for
+            accuracy calculation.
+        alignment (str, optional): method to align the prediction with the
+            groundtruth. Supported options are:
+            - ``'none'``: no alignment will be applied
+            - ``'scale'``: align in the least-square sense in scale
+            - ``'procrustes'``: align in the least-square sense in scale,
+                rotation and translation.
+    Returns:
+        tuple: A tuple containing joint position errors
+        - mpjpe (float|np.ndarray[N]): mean per-joint position error.
+        - p-mpjpe (float|np.ndarray[N]): mpjpe after rigid alignment with the
+            ground truth
+    """
+    assert mask.any()
+
+    if alignment == 'none':
+        pass
+    elif alignment == 'procrustes':
+        pred = np.stack([
+            compute_similarity_transform(pred_i, gt_i)
+            for pred_i, gt_i in zip(pred, gt)
+        ])
+    elif alignment == 'scale':
+        pred_dot_pred = np.einsum('nkc,nkc->n', pred, pred)
+        pred_dot_gt = np.einsum('nkc,nkc->n', pred, gt)
+        scale_factor = pred_dot_gt / pred_dot_pred
+        pred = pred * scale_factor[:, None, None]
+    else:
+        raise ValueError(f'Invalid value for alignment: {alignment}')
+
+    error = np.linalg.norm(pred - gt, ord=2, axis=-1)[mask].mean()
+
+    return error
+
+
+def keypoint_accel_error(joints_gt, joints_pred, vis=None):
+    """Computes acceleration error:
+
+    Note that for each frame that is not visible, three entries in the
+    acceleration error should be zero'd out.
+    Args:
+        joints_gt (Nx14x3).
+        joints_pred (Nx14x3).
+        vis (N).
+    Returns:
+        error_accel (N-2).
+    """
+    # (N-2)x14x3
+    accel_gt = joints_gt[:-2] - 2 * joints_gt[1:-1] + joints_gt[2:]
+    accel_pred = joints_pred[:-2] - 2 * joints_pred[1:-1] + joints_pred[2:]
+
+    normed = np.linalg.norm(accel_pred - accel_gt, axis=2)
+
+    if vis is None:
+        new_vis = np.ones(len(normed), dtype=bool)
+    else:
+        invis = np.logical_not(vis)
+        invis1 = np.roll(invis, -1)
+        invis2 = np.roll(invis, -2)
+        new_invis = np.logical_or(invis, np.logical_or(invis1, invis2))[:-2]
+        new_vis = np.logical_not(new_invis)
+
+    return np.mean(normed[new_vis], axis=1)
+
+
+def vertice_pve(pred_verts, target_verts=None):
+    """Computes per vertex error (PVE).
+
+    Args:
+        verts_gt (Nx6890x3).
+        verts_pred (Nx6890x3).
```

Comment:
If body_model is SMPL-X, then it is not 6890.

Reply:
Use `verts_num` instead